src/java_bytecode/java_bytecode_language.cpp

```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
+/// literal (i.e. the result of a Classname.class expression in Java). A
+/// constant global symbol will be created if it doesn't already exist.
+/// \param class_ref: the raw bytecode expression for the class literal
+/// \param symbol_table: global symbol table; may gain a new symbol
+/// \return symbol expression for the Class literal.
+static symbol_exprt get_or_create_class_literal_symbol(
+  const exprt &class_ref,
+  symbol_tablet &symbol_table)
+{
+  const irep_idt &class_id = class_ref.type().get(ID_identifier);
+  symbol_typet java_lang_Class("java::java.lang.Class");
+  symbol_exprt symbol_expr(
+    id2string(class_id)+"@class_model",
+    java_lang_Class);
+
+  if(!symbol_table.has_symbol(symbol_expr.get_identifier()))
+  {
+    symbolt new_class_symbol;
+    new_class_symbol.name = symbol_expr.get_identifier();
+    new_class_symbol.type = symbol_expr.type();
+    // name without "java::" prefix:
+    INVARIANT(
+      has_prefix(id2string(new_class_symbol.name), "java::"),
+      "class name should have java:: prefix");
+    new_class_symbol.base_name =
+      id2string(new_class_symbol.name).substr(6);
+    new_class_symbol.mode = ID_java;
+    new_class_symbol.is_lvalue = true;
+    new_class_symbol.is_state_var = true;
+    new_class_symbol.is_static_lifetime = true;
+    symbol_table.add(new_class_symbol);
+  }
+
+  return symbol_expr;
+}
+
+/// Creates global variables for constants mentioned in a given method. These
+/// are either string literals, or class literals (the java.lang.Class instance
+/// returned by `(some_reference_typed_expression).class`). The method parse
+/// tree is rewritten to directly reference these globals.
+/// \param parse_tree: parse tree to search for constant global references
+/// \param symbol_table: global symbol table, to which constant globals will be
+///   added.
+/// \param string_refinement_enabled: true if `--refine-stings` is active,
+///   which changes how string literals are structured.
+static void generate_constant_global_variables(
+  java_bytecode_parse_treet &parse_tree,
+  symbol_tablet &symbol_table,
+  bool string_refinement_enabled)
+{
+  for(auto &method : parse_tree.parsed_class.methods)
+  {
+    for(auto &instruction : method.instructions)
+    {
+      if(instruction.statement == "ldc" ||
+         instruction.statement == "ldc2" ||
+         instruction.statement == "ldc_w" ||
+         instruction.statement == "ldc2_w")
+      {
+        INVARIANT(
+          instruction.args.size() != 0,
+          "ldc instructions should have an argument");
+
+        // 1) Pushing a String causes a reference to a java.lang.String object
+        // to be constructed and pushed onto the operand stack.
+
+        // 2) Pushing an int or a float causes a primitive value to be pushed
+        // onto the stack.
+
+        // 3) Pushing a Class constant causes a reference to a java.lang.Class
+        // to be pushed onto the operand stack
```

Comment:
These comments describe "what" and not why, which renders them unnecessary.

Reply:
Not done; this describes the semantics of the opcode more clearly than reading the code.